tools/skbstat.py

```diff
@@ -0,0 +1,996 @@
+		if (skb_ip_match(skb, &filter)) {
+			return cb(ctx, args);
+		}
+	} else if ((filter.ip_version == 0 || filter.ip_version == 6)
+		   && skb_is_ipv6(skb)) {
+		if (skb_ip6_match(skb, &filter)) {
+			return cb(ctx, args);
+		}
+	}
+
+	return 0;
+}
+
+static inline int __trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+static inline int __trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_input(ctx, sk);
+}
+
+int trace_skb_input_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			 struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_output(ctx, sk);
+}
+
+int trace_skb_output_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			  struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_forward(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.forward = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_skb_drop(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.drop = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+
+int trace_skb_consume(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+"""
+
+# apply filters
+bpf_text = bpf_text.replace('SADDR_FILTER', saddr_filter)
+bpf_text = bpf_text.replace('DADDR_FILTER', daddr_filter)
+bpf_text = bpf_text.replace('SPORT_FILTER', sport_filter)
+bpf_text = bpf_text.replace('DPORT_FILTER', dport_filter)
+bpf_text = bpf_text.replace('PROTO_FILTER', proto_filter)
+bpf_text = bpf_text.replace('FLAGS_FILTER', flags_filter)
+bpf_text = bpf_text.replace('IP_VERSION_FILTER', ip_version_filter)
+
+# dump ebpf and exit
+if args.ebpf:
+    print(bpf_text)
+    exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+
+# probe list
+probe_list_input_dev = {
+    "netif_receive_skb_internal": ("trace_skb_input", 30),
```

Comment:
I hit the following error:
```
$ sudo ./skbstat.py --input
ERROR: netif_receive_skb_internal() kernel function not found or traceable. Older kernel versions not supported.
```

The function is inlined in netif_receive_skb
```
int netif_receive_skb(struct sk_buff *skb)
{       
        int ret;
        
        trace_netif_receive_skb_entry(skb);

        ret = netif_receive_skb_internal(skb);
        trace_netif_receive_skb_exit(ret);
        
        return ret;
}       
EXPORT_SYMBOL(netif_receive_skb);
```
Any particular reason you did not trace func `netif_receive_skb` or tracepoint/raw_tracepoint for netif_receive_skb_entry?

Reply:
Most of the driver code in kernel source tree, they call `napi_gro_receive` instead of `netif_receive_skb`. The backtrace of input routine on my machine showed like this:
```
                ip_protocol_deliver_rcu+0x36
                ip_local_deliver_finish+0x44
                ip_local_deliver+0x6b
                ip_rcv+0x52
                __netif_receive_skb_one_core+0x50
                netif_receive_skb_internal+0x2f
                napi_gro_receive+0x107
```

Reply:
Also, code with tracepoint and kprobe are so different and hard to write neatly. So this version of skbstat only contains kprobes, no tracepoints. Maybe next version will solve this issue.

Reply:
Another alternative is tracepoint net:napi_gro_receive_entry if you cares about the caller is napi_gro_receive. But not sure whether it covers all cases. Or you could use net:netif_receive_skb directly as the function __netif_receive_skb_core is prone to potential inlining problem.

Reply:
yes, tracepoint is always stable. I'll try to implement with tracepoint.